src/node_crypto.cc

```diff
@@ -2579,6 +2580,815 @@ static bool IsSupportedAuthenticatedMode(const EVP_CIPHER_CTX* ctx) {
+      CHECK_EQ(config.type_.ToChecked(), kKeyEncodingSEC1);
+      const unsigned char* p = reinterpret_cast<const unsigned char*>(key);
+      pkey.reset(d2i_PrivateKey(EVP_PKEY_EC, nullptr, &p, key_len));
+    }
+  }
+
+  // OpenSSL can fail to parse the key but still return a non-null pointer.
+  if (ERR_peek_error() != 0)
+    pkey.reset();
+
+  return pkey;
+}
+
+ByteSource::ByteSource(ByteSource&& other)
+      : data_(other.data_),
+        allocated_data_(other.allocated_data_),
+        size_(other.size_) {
+  other.allocated_data_ = nullptr;
+}
+
+ByteSource::~ByteSource() {
+  OPENSSL_clear_free(allocated_data_, size_);
+}
+
+ByteSource& ByteSource::operator=(ByteSource&& other) {
+  if (&other != this) {
+    OPENSSL_clear_free(allocated_data_, size_);
+    data_ = other.data_;
+    allocated_data_ = other.allocated_data_;
+    other.allocated_data_ = nullptr;
+    size_ = other.size_;
+  }
+  return *this;
+}
+
+const char* ByteSource::get() const {
+  return data_;
+}
+
+size_t ByteSource::size() const {
+  return size_;
+}
+
+ByteSource ByteSource::FromStringOrBuffer(Environment* env,
+                                          Local<Value> value) {
+  return Buffer::HasInstance(value) ? FromBuffer(value)
+                                    : FromString(env, value.As<String>());
+}
+
+ByteSource ByteSource::FromString(Environment* env, Local<String> str,
+                                  bool ntc) {
+  CHECK(str->IsString());
+  size_t size = str->Utf8Length(env->isolate());
+  size_t alloc_size = ntc ? size + 1 : size;
+  char* data = MallocOpenSSL<char>(alloc_size);
+  int opts = String::NO_OPTIONS;
+  if (!ntc) opts |= String::NO_NULL_TERMINATION;
+  str->WriteUtf8(env->isolate(), data, alloc_size, nullptr, opts);
+  return Allocated(data, size);
+}
+
+ByteSource ByteSource::FromBuffer(Local<Value> buffer, bool ntc) {
+  size_t size = Buffer::Length(buffer);
+  if (ntc) {
+    char* data = MallocOpenSSL<char>(size + 1);
+    memcpy(data, Buffer::Data(buffer), size);
+    data[size] = 0;
+    return Allocated(data, size);
+  }
+  return Foreign(Buffer::Data(buffer), size);
+}
+
+ByteSource ByteSource::NullTerminatedCopy(Environment* env,
+                                          Local<Value> value) {
+  return Buffer::HasInstance(value) ? FromBuffer(value, true)
+                                    : FromString(env, value.As<String>(), true);
+}
+
+ByteSource ByteSource::FromSymmetricKeyObject(Local<Value> handle) {
+  CHECK(handle->IsObject());
+  KeyObject* key = Unwrap<KeyObject>(handle.As<Object>());
+  CHECK(key);
+  // Note that this is only safe as long as the key object cannot be accessed
+  // by other threads.
+  return Foreign(key->GetSymmetricKey(), key->GetSymmetricKeySize());
+}
+
+ByteSource::ByteSource(const char* data, char* allocated_data, size_t size)
+      : data_(data),
+        allocated_data_(allocated_data),
+        size_(size) {}
+
+ByteSource ByteSource::Allocated(char* data, size_t size) {
```

Comment:
IMHO this is a leaky abstraction. I would suggest you make two classes `OwningByteSource` and `ForignByteSource`, that inherit from an abstract `ByteSource`

Reply:
If this is supposed to by a by-value only class, then you could implement this polymorphism with the [Pimpl pattern](https://en.cppreference.com/w/cpp/language/pimpl)

Reply:
I would prefer not to change this unless someone feels strongly about it (and we can still change it later). It is not that I like this version better than what you are proposing, but I feel like this is such a trivial abstraction that the added complexity would not necessarily improve things.

Reply:
If it's scope limited to `node::crypto` I agree we can defer improving this.